docs/development/test-guidelines.adoc

```diff
@@ -0,0 +1,209 @@
+    allows for easier writing.
+<3> Testing the exported `Say` function.
+<4> Marking the test for the `Say` function test as: "executable in parallel".
+<5> Testing the exported `Shout` function.
+<6> Marking the test for the `Shout` function test as: "executable in parallel".
+
+A few things to notice in <<lib-blackbox-test-dot-go-source>> is the package name: `bar_test`.
+This is an exception to the Go compiler because normally only package names with the same name are
+allowed in one folder. Go doesn't complain if you add `_test` to the name though.
+A nice side-effect of this exception is that it allows us to better determine the coverage of
+a package. If the test functions needed to live inside of a different directory, we would
+not have been able to collect coverage information because currently Go can only collect
+this for a single package or directory.
+
+NOTE: Have you noticed the `t.Parallel()` calls? This is a hint for `go test` so that it knows
+that the test `t` can be executed in parallel. Normally `go test` executes tests sequentially,
+but it also tries to executes as many tests in parallel as there are CPUs and as there are tests
+that have been marked to be executable in parallel. In this example, none of our tests
+has any side effects and so they can be executed in parallel.
+The number of parallel executable tests can be overwritten by the `-parallel` switch.
+
+Let's move on to our whitebox test file.
+
+[[lib-whitebox-test-dot-go-source]]
+[source,go]
+.$GOPATH/src/foo/bar/lib_whitebox_test.go
+----
+package bar // <1>
+
+import "testing"
+
+func TestToUpper(t *testing.T) { // <2>
+	t.Parallel() // <3>
+	if toUpper("hello") != "HELLO" {
+		t.Fail()
+	}
+}
+----
+<1> Notice the package name is just `bar` so we have access to all symbols (exported and unexported ones).
+<2> Test the unexported symbol `toUpper`.
+<3> Marking the test for the `toUpper` function test as: "executable in parallel".
+
+The whitebox test <<lib-whitebox-test-dot-go-source>> uses the `bar` package name
+and thereby gets access to all the internal of a package. There is also no need
+to import the `foo/bar` package. We can just use the `toUpper()` function.
+
+In order to confirm that our code actually works, let's jump to a shell and run these commands:
+
+[source,bash]
+.Run tests and collect coverage information
+----
+$ go test -v -coverprofile cov.out foo/bar # <1>
+=== RUN   TestToUpper
+--- PASS: TestToUpper (0.00s)
+=== RUN   TestSay
+=== RUN   TestShout
+--- PASS: TestSay (0.00s)
+--- PASS: TestShout (0.00s)
+PASS
+coverage: 100.0% of statements
+ok  	foo/bar	0.002s
+
+
+$ go tool cover -func=cov.out # <2>
+foo/bar/lib.go:9:	Say		100.0%
+foo/bar/lib.go:15:	Shout		100.0%
+foo/bar/lib.go:20:	toUpper		100.0%
+total:			(statements)	100.0%
+----
+<1> Runs all the tests in the `foo/bar` package and collects coverage information in the `cov.out` file.
+<2> Prints coverage information per function in the package.
```

Comment:
As discussed in chat, the coverage numbers for a package or function we generate from our makefile do not include coverage from test located in a different package. This will ignore coverage from end-to-end and cross-package integration tests.
As a group, we need to decide if that is a problem and needs to be fixed. There is a golang bug against this: https://github.com/golang/go/issues/6909 Later comments include pointers to tools that work around this.